src/V3SplitVar.cpp

```diff
@@ -0,0 +1,250 @@
+bool SplitVarVisitor::canSplit(const AstVar* nodep) {
+    if (AstNodeDType* dtypep = nodep->subDTypep()) {
+        const std::pair<uint32_t, uint32_t> dim = dtypep->dimensions(false);
+        UINFO(5, "Unpacked Dim of \"" << nodep->name() << "\":" << dim.second << '\n');
+        // Support just 1D in unpacked side.
+        // Traced or public variable cannot be splitted.
+        return dim.second == 1 && !nodep->isSigPublic() && !nodep->isTrace();
+    }
+    return false;
+}
+
+int SplitVarVisitor::debug(bool reset) {
+    static int level = -1;
+    if (VL_UNLIKELY(level < 0) || reset) {
+        level = v3Global.opt.debugSrcLevel(__FILE__);
+    }
+    return level;
+}
+
+void SplitVarVisitor::visit(AstNode* nodep) {
+    iterateChildren(nodep);
+}
+
+void SplitVarVisitor::visit(AstNodeModule* nodep) {
+    if (m_modp) {
+        SplitVarVisitor visitor(nodep);
+        visitor.split();
+    } else {
+        m_modp = nodep;
+        iterateChildren(nodep);
+    }
+}
+
+void SplitVarVisitor::visit(AstVar* nodep) {
+    m_lastVar = nodep;
+}
+
+void SplitVarVisitor::visit(AstPragma* pragmap) {
+    if (pragmap->pragType() != AstPragmaType::SPLIT_VAR) return;  // nothing to do
+    if (!m_lastVar) {
+        pragmap->v3warn(SPLITVAR, "Stray pragma of split_var is detected.");
+    } else if (!canSplit(m_lastVar)) {
+        pragmap->v3warn(SPLITVAR,
+                        "Pragma split_var is specified on a variable whose type is not supported yet. "
+                        "Only unpacked 1D array is supported by this version.");
+    } else {  // finally find a good candidate
+        UINFO(4, m_lastVar->name() << " is added to candidate list.\n");
+        m_refs.insert(std::make_pair(m_lastVar, std::vector<AstArraySel*>()));
+    }
+    m_lastVar = NULL;
+    pragmap->unlinkFrBack()->deleteTree();  // consume the pragma here anyway.
+}
+
+void SplitVarVisitor::visit(AstArraySel* nodep) {
+    AstVarRef* const vrefp = VN_CAST(nodep->fromp(), VarRef);
+    if (!vrefp) return;
+
+    AstVar* const varp = vrefp->varp();
+    if (m_refs.find(varp) == m_refs.end()) return;  // variable without split_var pragma
+
+    AstConst* indexp = VN_CAST(nodep->bitp(), Const);
+    if (!indexp) {  // index is not constant yet.
+        UINFO(4, "Index of " << vrefp->name() << " is " << nodep->bitp()
+                             << " which is not constant.\n");
+        AstNode* const constified = V3Const::constifyEdit(nodep->bitp());
+        UINFO(4, "After constified:" << constified << '\n');
+        indexp = VN_CAST(constified, Const);
+    }
+
+    if (indexp) {  // OK
+        m_refs[varp].push_back(nodep);
+    } else {
+        nodep->bitp()->v3warn(SPLITVAR,
+                              "Variable \"" << vrefp->name() << "\" will not be splitted "
+                              "because index cannot be determined statically.");
+        if (m_refs.find(varp) != m_refs.end()) m_refs.erase(varp);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redundant checking of STL container element existence before removing it.](https://app.codacy.com/app/wsnyder/verilator/pullRequest?prid=4724658)